File: src/features/typing/data/java/letter_combinations_of_a_phone_number.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private static final String[] DIGIT_LETTERS = {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
    };

    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits.isEmpty()) {
            return result;
        }
        backtrack(digits, 0, new StringBuilder(), result);
        return result;
    }

    private void backtrack(String digits, int index, StringBuilder current, List<String> result) {
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }
        String letters = DIGIT_LETTERS[digits.charAt(index) - '0'];
        for (char ch : letters.toCharArray()) {
            current.append(ch);
            backtrack(digits, index + 1, current, result);
            current.deleteCharAt(current.length() - 1);
        }
    }
}

File: src/features/typing/data/dart/letter_combinations_of_a_phone_number.dart
class Solution {
  static const _digitLetters = {
    '2': 'abc',
    '3': 'def',
    '4': 'ghi',
    '5': 'jkl',
    '6': 'mno',
    '7': 'pqrs',
    '8': 'tuv',
    '9': 'wxyz',
  };

  List<String> letterCombinations(String digits) {
    final result = <String>[];
    if (digits.isEmpty) {
      return result;
    }
    void backtrack(int index, String current) {
      if (index == digits.length) {
        result.add(current);
        return;
      }
      final letters = _digitLetters[digits[index]] ?? '';
      for (final ch in letters.split('')) {
        backtrack(index + 1, current + ch);
      }
    }
    backtrack(0, '');
    return result;
  }
}
